crates/sui-graphql-rpc/src/context_data/db_data_provider.rs

```diff
@@ -2123,6 +2121,114 @@ pub(crate) fn validate_cursor_pagination(
     Ok(())
 }
 
+pub(crate) fn convert_to_validators(
+    validators: Vec<SuiValidatorSummary>,
+    system_state: Option<&NativeSuiSystemStateSummary>,
+) -> Vec<Validator> {
+    validators
+        .iter()
+        .map(|v| {
+            let at_risk = system_state
+                .and_then(|system_state| {
+                    system_state
+                        .at_risk_validators
+                        .iter()
+                        .find(|&(address, _)| address == &v.sui_address)
+                })
+                .map(|&(_, value)| value);
+
+            let report_records = system_state
+                .and_then(|system_state| {
+                    system_state
+                        .validator_report_records
+                        .iter()
+                        .find(|&(address, _)| address == &v.sui_address)
+                })
+                .map(|(_, value)| {
+                    value
+                        .iter()
+                        .map(|address| SuiAddress::from_array(address.to_inner()))
+                        .collect::<Vec<_>>()
+                });
+
+            let credentials = ValidatorCredentials {
+                protocol_pub_key: Some(Base64::from(v.protocol_pubkey_bytes.clone())),
+                network_pub_key: Some(Base64::from(v.network_pubkey_bytes.clone())),
+                worker_pub_key: Some(Base64::from(v.worker_pubkey_bytes.clone())),
+                proof_of_possession: Some(Base64::from(v.proof_of_possession_bytes.clone())),
+                net_address: Some(v.net_address.clone()),
+                p2p_address: Some(v.p2p_address.clone()),
+                primary_address: Some(v.primary_address.clone()),
+                worker_address: Some(v.worker_address.clone()),
+            };
+            Validator {
+                address: Address {
+                    address: SuiAddress::from(v.sui_address),
+                },
+                next_epoch_credentials: Some(credentials.clone()),
+                credentials: Some(credentials),
+                name: Some(v.name.clone()),
+                description: Some(v.description.clone()),
+                image_url: Some(v.image_url.clone()),
+                project_url: Some(v.project_url.clone()),
+
+                operation_cap_id: SuiAddress::from_array(**v.operation_cap_id),
+                staking_pool_id: SuiAddress::from_array(**v.staking_pool_id),
+                exchange_rates_id: SuiAddress::from_array(**v.exchange_rates_id),
```

Comment:
Don't these panic if the array length is wrong?  If so, we should wrap them in internal errors.

Reply:
Indeed pre-existing, but the underlying data is size 32, and `from_array` is expecting [u8; 32] so it would fail at compile time otherwise. It just copies bytes and doesn't panic either.

Reply:
Oh I see -- these are ObjectIDs.  There are `From` implementation for those on GraphQL's `SuiAddress` type:

https://github.com/MystenLabs/sui/blob/5c1276057b91331450ff37db5f47c72fb3e28bf0/crates/sui-graphql-rpc/src/types/sui_address.rs#L85-L96